clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -614,13 +618,28 @@ private void handleProduceResponse(ClientResponse response, Map<TopicPartition,
                                 .collect(Collectors.toList()),
                             p.errorMessage());
                     ProducerBatch batch = batches.get(tp);
-                    completeBatch(batch, partResp, correlationId, now);
+                    completeBatch(batch, partResp, correlationId, now, partitionsWithUpdatedLeaderInfo);
+
+                    final short requestVersion = response.requestHeader().apiVersion();
+                    if (requestVersion >= 16 && !partitionsWithUpdatedLeaderInfo.isEmpty()) {
+                        List<Node> leaderNodes = produceResponse.data().nodeEndpoints().stream()
+                            .map(e -> new Node(e.nodeId(), e.host(), e.port(), e.rack()))
+                            .filter(e -> !e.equals(Node.noNode()))
+                            .collect(
+                                Collectors.toList());
+                        Set<TopicPartition> updatedPartitions = metadata.updatePartially(partitionsWithUpdatedLeaderInfo, leaderNodes);
+                        if (log.isTraceEnabled()) {
+                            updatedPartitions.forEach(
+                                part -> log.trace("For {} leader was updated.", part)
+                            );
+                        }
+                    }
```

Comment:
Is this worth the effort to generalize and share among `Sender` and `AbstractFetch`?

Reply:
No, as there are subtle differences as in different response types & Fetcher has extra actions after leader is updated. So the common part is really is just 1-2 statements, so not worth it IMO.